File: packages/htf-js/htf/features.js
// SPDX-License-Identifier: Apache-2.0
// Copyright 2025-2026 ZHU Shengling
((global) => {
  "use strict";

  const HTF = global.HTF || (global.HTF = {});
  const FeatureModule = HTF.FeatureModule;

  const isNumber = (value) => typeof value === "number" && Number.isFinite(value);

  const toFloatList = (seq) => {
    const out = [];
    seq.forEach((val) => {
      if (isNumber(val)) {
        out.push(val);
      }
    });
    return out;
  };

  const mean = (values) => {
    if (!values.length) {
      return null;
    }
    return values.reduce((sum, v) => sum + v, 0) / values.length;
  };

  class SingleFieldStatsFeature extends FeatureModule {
    constructor({ field_name, prefix } = {}) {
      super();
      this.field_name = field_name;
      this.prefix = prefix;
    }

    compute(window) {
      const vals = toFloatList(window.map((rec) => (rec ? rec[this.field_name] : undefined)));
      if (!vals.length) {
        return {
          [`${this.prefix}_count`]: 0,
          [`${this.prefix}_mean`]: null,
          [`${this.prefix}_min`]: null,
          [`${this.prefix}_max`]: null,
        };
      }
      return {
        [`${this.prefix}_count`]: vals.length,
        [`${this.prefix}_mean`]: mean(vals),
        [`${this.prefix}_min`]: Math.min(...vals),
        [`${this.prefix}_max`]: Math.max(...vals),
      };
    }
  }

  class LastRecordEchoFeature extends FeatureModule {
    constructor({ fields } = {}) {
      super();
      this.fields = Array.isArray(fields) ? fields : [];
    }

    compute(window) {
      if (!window.length) {
        const empty = {};
        this.fields.forEach((name) => {
          empty[name] = null;
        });
        return empty;
      }
      const last = window[window.length - 1] || {};
      const out = {};
      this.fields.forEach((name) => {
        out[name] = Object.hasOwn(last, name) ? last[name] : null;
      });
      return out;
    }
  }

  HTF.SingleFieldStatsFeature = SingleFieldStatsFeature;
  HTF.LastRecordEchoFeature = LastRecordEchoFeature;
})(typeof window !== "undefined" ? window : globalThis);

File: packages/htf-py/htf/features.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2025-2026 ZHU Shengling
from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from typing import Any

from .timeframe import FeatureDict, FeatureModule, Record


def _is_number(x: Any) -> bool:
    return isinstance(x, (int, float)) and not isinstance(x, bool)


def _to_float_list(seq: Sequence[Any]) -> list[float]:
    out: list[float] = []
    for v in seq:
        if _is_number(v):
            out.append(float(v))
    return out


def _mean(values: Sequence[float]) -> float | None:
    if not values:
        return None
    return float(sum(values) / len(values))


@dataclass
class SingleFieldStatsFeature(FeatureModule):
    field_name: str
    prefix: str

    def compute(self, window: Sequence[Record]) -> FeatureDict:
        vals = _to_float_list([rec.get(self.field_name) for rec in window])
        if not vals:
            return {
                f"{self.prefix}_count": 0,
                f"{self.prefix}_mean": None,
                f"{self.prefix}_min": None,
                f"{self.prefix}_max": None,
            }

        return {
            f"{self.prefix}_count": len(vals),
            f"{self.prefix}_mean": _mean(vals),
            f"{self.prefix}_min": min(vals),
            f"{self.prefix}_max": max(vals),
        }


@dataclass
class LastRecordEchoFeature(FeatureModule):
    fields: Sequence[str]

    def compute(self, window: Sequence[Record]) -> FeatureDict:
        if not window:
            return {name: None for name in self.fields}
        last = window[-1]
        return {name: last.get(name) for name in self.fields}
